### PR TITLE
api dependency reduction - utils constants

### PR DESCRIPTION
**Pull request by Joerger**
Changes:
 - Moved necessary functions and constants out of `/teleport` and `/lib/utils` and into `/api/constants` and `/api/utils`.
 - Moved a lot of business logic related functions and methods out of `/api/types` and into `/lib/services`, rather than moving the utils functions they use. 
   - One goal of the api refactoring is decouple business logic from the `api`, but this would be a large PR, so instead it'll be done incrementally as these functions and files are touched for other reasons.
 - Removed `logrus` and `lib/fixtures` dependencies in `api` test files.

### CODE REVIEW

File: api/constants/constants.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2020 Gravitational, Inc.
```
Comment (a-palchikov) [inline api/constants/constants.go:2]:
```suggestion
Copyright 2021 Gravitational, Inc.
```

File: api/types/license_test.go [hunk 2]
```diff
@@ -98,12 +92,16 @@ func (s *LicenseSuite) TestUnmarshal(c *check.C) {
 		out, err := UnmarshalLicense([]byte(tc.input))
 		if tc.err == nil {
 			c.Assert(err, check.IsNil, comment)
-			fixtures.DeepCompare(c, tc.expected, out)
+			if !reflect.DeepEqual(tc.expected, out) {
```
Comment (a-palchikov) [inline api/types/license_test.go:95]: gocheck has assertions for this:
```go
c.Assert(out, check.DeepEquals, tc.expected, check.Commentf(tc.description))
```

File: api/types/server.go [hunk 3]
```diff
@@ -23,9 +23,9 @@ import (
 	"strings"
 	"time"
 
-	"github.com/gravitational/teleport/lib/utils"
-
 	"github.com/gogo/protobuf/proto"
+	"github.com/gravitational/teleport/api/utils"
```
Comment (a-palchikov) [inline api/types/server.go:27]:
```go
"strings"
"time"

"github.com/gravitational/teleport/api/utils"

"github.com/gogo/protobuf/proto"
"github.com/gravitational/trace"
```

File: api/types/teleport_role.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2020 Gravitational, Inc.
```
Comment (a-palchikov) [inline api/types/teleport_role.go:2]:
```suggestion
Copyright 2021 Gravitational, Inc.
```

File: api/types/teleport_role.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	RoleProxy TeleportRole = "Proxy"
+	// RoleAdmin is admin role
+	RoleAdmin TeleportRole = "Admin"
+	// RoleProvisionToken is a role for nodes authenticated using provisioning tokens
+	RoleProvisionToken TeleportRole = "ProvisionToken"
+	// RoleTrustedCluster is a role needed for tokens used to add trusted clusters.
+	RoleTrustedCluster TeleportRole = "Trusted_cluster"
+	// RoleSignup is for first time signing up users
+	RoleSignup TeleportRole = "Signup"
+	// RoleNop is used for actions that already using external authz mechanisms
```
Comment (a-palchikov) [inline api/types/teleport_role.go:51]:
```suggestion
	// RoleNop is used for actions that are already using external authz mechanisms
```

File: api/utils/json.go [hunk 6]
```diff
@@ -17,12 +17,41 @@ limitations under the License.
...
+	"unicode"
 
+	"github.com/ghodss/yaml"
 	"github.com/gravitational/configure/jsonschema"
 	"github.com/gravitational/trace"
+	jsoniter "github.com/json-iterator/go"
 )
 
+// FastUnmarshal uses the json-iterator library for fast JSON unmarshalling.
+// Note, this function marshals floats with 6 digits precision.
```
Comment (a-palchikov) [inline api/utils/json.go:31]:
```suggestion
// Note, this function unmarshals floats with 6 digits precision.
```
?

File: lib/services/role.go [hunk 7]
```diff
@@ -493,6 +493,50 @@ func applyValueTraits(val string, traits map[string][]string) ([]string, error)
 	return interpolated, nil
 }
 
+// score is a sorting score of the rule, the larger the score, the more
```
Comment (a-palchikov) [inline lib/services/role.go:496]:
```suggestion
// scoreRule is a sorting score of the rule, the larger the score, the more
```
I'd actually name it - `ruleScore` as it's the rule's score and not the other way around

File: roles.go [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2015 Gravitational, Inc.
+Copyright 2020 Gravitational, Inc.
```
Comment (a-palchikov) [inline roles.go:2]:
```suggestion
Copyright 2021 Gravitational, Inc.
```

File: api/types/resetpasswordtoken_test.go [hunk 9]
```diff
@@ -75,15 +74,11 @@ func (s *ResetPasswordTokenSuite) TestUnmarshal(c *check.C) {
 		comment := check.Commentf("test case %q", tc.description)
 		out, err := marshaler.Unmarshal([]byte(tc.input))
 		c.Assert(err, check.IsNil, comment)
-		if !reflect.DeepEqual(tc.expected, out) {
-			c.Fatalf("expected %v, but got %v", tc.expected, out)
-		}
+		c.Assert(out, check.DeepEquals, tc.expected, check.Commentf(tc.description))
```
Comment (a-palchikov) [inline api/types/resetpasswordtoken_test.go:77]: nit: I did not notice it before, but there's already a `comment` (value of type check.Commentf) which you can use here:
```go
c.Assert(out, check.DeepEquals, tc.expected, comment)
```

**Review by a-palchikov — APPROVED** [review]

File: api/types/role.go [hunk 10]
```diff
@@ -680,50 +680,6 @@ func (r *Rule) CheckAndSetDefaults() error {
...
-//
-// * nRule matching wildcard resource is less specific
-// than same rule matching specific resource.
-// * Rule that has wildcard verbs is less specific
-// than the same rules matching specific verb.
-// * Rule that has where section is more specific
-// than the same rule without where section.
-// * Rule that has actions list is more specific than
-// rule without actions list.
-func (r *Rule) IsMoreSpecificThan(o Rule) bool {
```
Comment (r0mant) [inline api/types/role.go:723]: Curious why did this (along with `score()` above) move back to lib/services?

File: api/types/teleport_role.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// TeleportRole identifies the role of an SSH connection. Unlike "user roles"
+// introduced as part of RBAC in Teleport 1.4+ these are built-in roles used
+// for different Teleport components when connecting to each other.
+type TeleportRole string
```
Comment (r0mant) [inline api/types/teleport_role.go:29]: Any reason these aren't named just `Role` and `Roles` respectively?

File: api/types/tunnel.go [hunk 12]
```diff
@@ -179,40 +180,23 @@ func (r *ReverseTunnelV2) Check() error {
...
-	for _, addr := range r.Spec.DialAddrs {
-		if _, err := utils.ParseAddr(addr); err != nil {
+	for _, addr := range r.GetDialAddrs() {
+		if addr == "" {
+			return trace.BadParameter("missing parameter address")
+		}
+		if !strings.Contains(addr, "://") {
+			return nil
+		}
+		if _, err := url.Parse(addr); err != nil {
```
Comment (r0mant) [inline api/types/tunnel.go:193]: Something that's a little confusing to me is, how do we determine now what kinds of checks are done in api/ on the client and which are on the server? For example, here you're parsing the URL, but somewhere above (I think it was OIDC or other connector) url.Parse checks were removed. I think we need to be consistent.

File: api/types/wrappers/wrappers_test.go [hunk 13]
```diff
@@ -33,10 +31,6 @@ var _ = check.Suite(&WrappersSuite{})
 
 func TestWrappers(t *testing.T) { check.TestingT(t) }
 
-func (s *WrappersSuite) SetUpSuite(c *check.C) {
-	utils.InitLoggerForTests(testing.Verbose())
```
Comment (r0mant) [inline api/types/wrappers/wrappers_test.go:37]: Same q about tests.

File: api/utils/json.go [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2015 Gravitational, Inc.
+Copyright 2021 Gravitational, Inc.
```
Comment (r0mant) [inline api/utils/json.go:2]: Don't really want to be nitpicking on the license headers, but I think we usually append the last year, not replace it so `2015-2021`. Noticed the same somewhere above as well.

File: api/utils/utils.go [hunk 15]
```diff
@@ -0,0 +1,37 @@
+/*
```
Comment (r0mant) [inline api/utils/utils.go:1]: Nit: I think something like `bool.go` would be more appropriate name for this file. Unless you intend to turn it into a collection of random utility methods later.

File: roles.go [hunk 16]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2015 Gravitational, Inc.
+Copyright 2021 Gravitational, Inc.
```
Comment (r0mant) [inline roles.go:2]: 2015-2021

File: api/constants/constants.go [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+const (
+	// MaxLeases serves as an identifying error string indicating that the
+	// semaphore system is rejecting an acquisition attempt due to max
+	// leases having already been reached.
+	MaxLeases = "err-max-leases"
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/constants/constants.go:98]: Yeah, I maintained their original groupings, but I can but it all in one const block instead.

File: api/types/github_test.go [hunk 18]
```diff
@@ -17,21 +17,13 @@ limitations under the License.
...
-
 	check "gopkg.in/check.v1"
 )
 
 type GithubSuite struct{}
 
 var _ = check.Suite(&GithubSuite{})
 
-func (s *GithubSuite) SetUpSuite(c *check.C) {
-	utils.InitLoggerForTests(testing.Verbose())
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/github_test.go:32]: This sets up the logger with our custom preferences with logrus. This was removed to remove the dependency on `utils` and `logrus`.

[the remark this answers is not in the document]
Comment (r0mant) [reply api/types/github_test.go:32]: Yes, but without this, if you run `go test -v ...`, it won't be showing all the logs anymore, will it?

[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/github_test.go:32]: I see what you mean now, I think we can move these back to services then. In fact #5384 moves most of these back anyways, so I'll update both PRs.

File: api/utils/utils.go (hunk 15 above)
Comment (Joerger) [reply api/utils/utils.go:1]: That was my original intention, but I don't think that'll happen after all.

File: api/utils/json.go (hunk 14 above)
Comment (Joerger) [reply api/utils/json.go:2]: Didn't know this, thanks!

File: api/types/teleport_role.go (hunk 11 above)
Comment (Joerger) [reply api/types/teleport_role.go:29]: This would conflict with the RBAC `types.Role`, so I thought `TeleportRole` made sense. We could also name it SSHRole or something different.

File: api/types/role.go (hunk 10 above)
Comment (Joerger) [reply api/types/role.go:723]: This function (and others) is business logic that shouldn't be in `api/types`. Since these PRs are trying to reduce dependencies and reduce the overall size of `api`, I moved this back. I should have made this a separate PR though.

File: api/types/tunnel.go (hunk 12 above)
Comment (Joerger) [reply api/types/tunnel.go:193]: In my opinion the only checks that should be done in the client should be basic value checks, so I should have moved this to `services` (will now). I want to completely clean up the validation logic across all types in a separate PR to make the separation between client and server validation very clear.

File: api/utils/json.go [hunk 19]
```diff
@@ -17,12 +17,41 @@ limitations under the License.
...
 
 import (
+	"bytes"
 	"encoding/json"
+	"unicode"
 
+	"github.com/ghodss/yaml"
 	"github.com/gravitational/configure/jsonschema"
 	"github.com/gravitational/trace"
+	jsoniter "github.com/json-iterator/go"
```
Comment (russjones) [inline api/utils/json.go:27]: Will these be moved out in another PR, seem to be going in the opposite direction we want (adding more dependencies).

File: api/utils/replace.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+	}
+	// if there is no match, return NotFound error
+	index := expr.FindAllStringIndex(input, -1)
+	if len(index) == 0 {
+		return "", trace.NotFound("no match found")
+	}
+	return expr.ReplaceAllString(input, replaceWith), nil
+}
+
+var replaceWildcard = regexp.MustCompile(`(\\\*)`)
```
Comment (russjones) [inline api/utils/replace.go:59]: Are these used in the `api` package?

File: api/utils/json.go (hunk 19 above)
Comment (Joerger) [reply api/utils/json.go:27]: Yes, these are actually moved back by #5384

File: api/utils/replace.go (hunk 20 above)
Comment (Joerger) [reply api/utils/replace.go:59]: Yes, it's used by `RoleMap`, which is pretty rooted in there. I can give another shot at moving it to `services` though.

File: lib/services/trustedcluster.go [hunk 21]
```diff
@@ -43,5 +47,98 @@ func ValidateTrustedCluster(tc TrustedCluster) error {
...
 	return nil
 }
+
+// ValidateRoleMap checks RoleMap for errors
+func ValidateRoleMap(r RoleMap) error {
+	_, err := parseRoleMap(r)
+	return trace.Wrap(err)
+}
+
+// RoleMapToString prints user friendly representation of role mapping
```
Comment (russjones) [inline lib/services/trustedcluster.go:63]: I don't think this is being called by anyone? You can drop it then move body of `parseRoleMap` into `ValidateRoleMap`.

File: api/types/teleport_role.go (hunk 11 above)
Comment (russjones) [reply api/types/teleport_role.go:29]: I'm not a huge fan of the name either, but can't think of anything better. @r0mant any ideas?

Comment (Joerger) [reply api/types/teleport_role.go:29]: Would `ClusterRole` be better? or maybe it's the "Role" part that is making it awkward?

Comment (r0mant) [reply api/types/teleport_role.go:29]: These are basically names of different Teleport components and services like auth server, proxy server, app server, etc., right? How about `SystemRole` or `ServiceRole`?

**Review by russjones — APPROVED** [review]
Looks good to me with role name change.
